File: src/stride_core/registry.py
"""Provider registry — dispatches a request to the right adapter for a user.

Per-user provider model: every user is bound to exactly one watch provider,
recorded in `data/{user_id}/config.json` as a top-level `provider` field.
Legacy users (config.json missing or no provider field) default to `'coros'`,
which is correct for all existing data — see DB migration v1's matching default.

The registry is constructed once at composition root (stride_server/main.py),
populated with one adapter per supported provider, and stored in
`app.state.registry`. Routes look up `for_user(user_id)` via a FastAPI
dependency to get the right adapter without ever importing a specific one.

Adding a new provider (Garmin/Polar/Suunto/...): build a new adapter that
implements DataSource, register it at composition root, done — no core or
route code changes.
"""

from __future__ import annotations

import json
from collections.abc import Iterable
from pathlib import Path

from .source import DataSource, ProviderInfo


# Default provider for users with no explicit setting. Matches the
# `provider TEXT NOT NULL DEFAULT 'coros'` SQL default in db.py — both are
# tied together by the assumption that all pre-multi-provider data is COROS.
DEFAULT_PROVIDER = "coros"


class UnknownProvider(KeyError):
    """Raised when ProviderRegistry.get() is called with an unregistered name."""

    def __init__(self, name: str) -> None:
        super().__init__(name)
        self.name = name

    def __str__(self) -> str:
        return f"No adapter registered for provider {self.name!r}"


class ProviderRegistry:
    """In-process registry of available DataSource adapters.

    Construct one at app boot, register all known adapters, hand to
    `create_app`. `for_user(user_id)` resolves the user's configured provider
    (default `'coros'`) and returns the matching adapter.
    """

    def __init__(self) -> None:
        self._sources: dict[str, DataSource] = {}
        self._default: str | None = None

    def register(self, source: DataSource, *, default: bool = False) -> None:
        """Add an adapter. If `default=True` (or this is the first registration),
        the adapter becomes the fallback for users without an explicit provider
        setting — which currently means *everyone*, since onboarding doesn't
        write the field yet."""
        name = source.info.name
        if name in self._sources:
            raise ValueError(f"Provider {name!r} already registered")
        self._sources[name] = source
        if default or self._default is None:
            self._default = name

    def get(self, name: str) -> DataSource:
        if name not in self._sources:
            raise UnknownProvider(name)
        return self._sources[name]

    def for_user(self, user: str) -> DataSource:
        """Resolve the adapter the given user is bound to.

        Reads the user's `provider` field from config.json; falls back to the
        registry's default (typically the first registered adapter) if the
        user has no setting. Raises `UnknownProvider` if the resolved name
        isn't registered (e.g. user's config references a provider this
        deployment doesn't support).
        """
        provider = read_user_provider(user, default=self._default or DEFAULT_PROVIDER)
        return self.get(provider)

    def names(self) -> Iterable[str]:
        return self._sources.keys()

    def all_infos(self) -> list[ProviderInfo]:
        return [src.info for src in self._sources.values()]

    def default_name(self) -> str | None:
        return self._default

    def __contains__(self, name: str) -> bool:
        return name in self._sources

    def __len__(self) -> int:
        return len(self._sources)


# ─────────────────────────────────────────────────────────────────────────────
# config.json provider field — read/write helpers
# ─────────────────────────────────────────────────────────────────────────────


def _user_config_path(user: str, base_dir: Path | None = None) -> Path:
    # Imported lazily so this module stays importable without forcing the
    # `data/` directory layout into stride_core's surface area.
    from .db import USER_DATA_DIR
    return (base_dir or USER_DATA_DIR) / user / "config.json"


def read_user_provider(
    user: str,
    *,
    default: str = DEFAULT_PROVIDER,
    base_dir: Path | None = None,
) -> str:
    """Resolve the watch provider for a user.

    Returns the `provider` field from `data/{user}/config.json`, or `default`
    if:
      - config.json is missing (e.g. user freshly created, not onboarded yet)
      - config.json is malformed (corrupt JSON, not a dict)
      - config.json has no `provider` field (legacy users — pre-multi-provider)

    Never raises on file/JSON errors; returning the default is the safer
    behavior for a sync request where forcing the user through re-onboarding
    would be worse than just trying COROS.
    """
    path = _user_config_path(user, base_dir)
    if not path.exists():
        return default
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return default
    if not isinstance(data, dict):
        return default
    value = data.get("provider")
    return str(value) if value else default


def write_user_provider(
    user: str,
    provider: str,
    *,
    base_dir: Path | None = None,
) -> None:
    """Persist a user's provider preference, preserving any other fields.

    config.json is shared with adapter-specific credentials (COROS stores
    `email` / `pwd_hash` / `access_token` here today). This helper only
    touches the `provider` key; everything else round-trips unchanged.
    Creates parent directories and the file itself if missing.
    """
    path = _user_config_path(user, base_dir)
    path.parent.mkdir(parents=True, exist_ok=True)
    if path.exists():
        try:
            data = json.loads(path.read_text(encoding="utf-8"))
            if not isinstance(data, dict):
                data = {}
        except (OSError, json.JSONDecodeError):
            data = {}
    else:
        data = {}
    data["provider"] = provider
    path.write_text(
        json.dumps(data, indent=2, ensure_ascii=False),
        encoding="utf-8",
    )

File: src/go/internal/registry/registry.go
// Package registry is the wiring layer that resolves which watch data source
// (provider) a user is bound to and constructs the concrete adapter. It sits
// ABOVE the provider contract and the concrete adapters, so it may import both
// coros and garmin without an import cycle (adapters never import registry).
//
// Provider binding source of truth (ADR 0010): the per-user binding is read from
// the local data/<uid>/config.json `provider` field for now, and will migrate to
// provider_credentials.provider (MySQL) later. Legacy users with no binding
// default to coros. login writes the binding via WriteProviderName so a
// subsequent flag-less sync resolves the same source.
package registry

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"time"

	"github.com/zhaochy1990/stride/internal/provider"
	"github.com/zhaochy1990/stride/internal/provider/coros"
	"github.com/zhaochy1990/stride/internal/provider/garmin"
	"github.com/zhaochy1990/stride/internal/storage"
)

// DefaultProvider is the binding for users with no explicit setting — matches the
// Python registry and the `provider DEFAULT 'coros'` SQL default (all
// pre-multi-provider data is COROS).
const DefaultProvider = "coros"

// Supported reports whether name is a known provider the binary can construct.
func Supported(name string) bool {
	return name == "coros" || name == "garmin"
}

// ProviderName returns the provider a user is bound to, read from
// data/<uid>/config.json. A missing file or absent `provider` field resolves to
// DefaultProvider; a malformed file is an error (fail loud rather than silently
// syncing the wrong source).
func ProviderName(dataDir, user string) (string, error) {
	raw, err := os.ReadFile(configPath(dataDir, user))
	if os.IsNotExist(err) {
		return DefaultProvider, nil
	}
	if err != nil {
		return "", fmt.Errorf("registry: read config for %s: %w", user, err)
	}
	var cfg map[string]any
	if err := json.Unmarshal(raw, &cfg); err != nil {
		return "", fmt.Errorf("registry: parse config for %s: %w", user, err)
	}
	if p, ok := cfg["provider"].(string); ok && p != "" {
		if !Supported(p) {
			return "", fmt.Errorf("registry: unknown provider %q bound to user %s", p, user)
		}
		return p, nil
	}
	return DefaultProvider, nil
}

// WriteProviderName persists the user's provider binding into config.json,
// preserving any existing fields (Python coros users keep email/pwd_hash/etc. in
// the same file). Creates the file/dir if absent.
func WriteProviderName(dataDir, user, providerName string) error {
	if !Supported(providerName) {
		return fmt.Errorf("registry: refusing to bind unknown provider %q", providerName)
	}
	path := configPath(dataDir, user)
	cfg := map[string]any{}
	if raw, err := os.ReadFile(path); err == nil {
		if err := json.Unmarshal(raw, &cfg); err != nil {
			return fmt.Errorf("registry: parse config for %s: %w", user, err)
		}
	} else if !os.IsNotExist(err) {
		return fmt.Errorf("registry: read config for %s: %w", user, err)
	}
	cfg["provider"] = providerName
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return err
	}
	out, err := json.MarshalIndent(cfg, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(path, out, 0o644)
}

// Build constructs the concrete provider adapter for providerName, wired to the
// store for both watch-data writes and the provider_credentials credential store.
func Build(providerName string, store *storage.Store, delay time.Duration) (provider.Provider, error) {
	switch providerName {
	case "coros":
		return coros.New(store, coros.NewStorageCredentialStore(store),
			coros.WithProviderRequestDelay(delay)), nil
	case "garmin":
		return garmin.New(store, garmin.NewStorageCredentialStore(store),
			garmin.WithProviderRequestDelay(delay)), nil
	default:
		return nil, fmt.Errorf("registry: unknown provider %q", providerName)
	}
}

func configPath(dataDir, user string) string {
	return filepath.Join(dataDir, user, "config.json")
}
